### ISSUE TITLE
De-duplication: clobber or merge?

### ISSUE THREAD

**Issue by steveluscher** [kissmetrics/km-support#21]
What's the behavior of KISSMetrics when it detects a duplicate event; will one clobber the other, or will you merge metadata from both?

For instance, say I send an event from the browser using the JS API, but send the same event with the same params/timestamp from the server. Will one clobber the other, or will I get the browser-y metadata from the JS API tracking call merged with the server tracking call?

Filing this here, because the documentation on de-duping doesn't cover this.

**Comment by nateklaiber** [comment]
Hey @steveluscher, thanks for the question.

I have passed this to our support engineer, @eskfung, who can get you some more information. 

I agree we need to have this information documented in support.

**Comment by eskfung** [comment]
@nateklaiber I'm not sure I even know which event takes precedence. I haven't encountered it too often and the one instance I can remember, it seemed to be randomly choosing which event to present.

**Comment by steveluscher** [comment]
You should merge the options. I'd like to be able to fire the same event on the client and the server; if the client request makes it out, great – we get a bunch of client-y data. If it fails (or is blocked), the server request acts as a last line of defense.

**Comment by eskfung** [comment]
@steveluscher now that I'm re-reading your original question, I think I'll need to use an example:

I'm not sure what browser metadata you're looking to save. Let's say you had JavaScript in place that looked like this:

``` javascript
_kmq.push(['identify', 'User123']);
_kmq.push(['record', 'An Event Happened', {'Property 1': 'Value 1', '_t':1379025365, '_d':1}]);
// Setting the _t and _d parameters here establish what time the event happened.
// http://support.kissmetrics.com/apis/specifications
```

Upon execution, we would interpret this as:

```
User123 did the event "An Event Happened", at Thu, 12 Sep 2013 22:36:05 GMT (translating 1379025365)
User123 got the value "Value 1" for the property "Property 1" at Thu, 12 Sep 2013 22:36:05 GMT
```

And that's _all we process from a JavaScript API call_.

### Example 1

Let's say you have Ruby backing your JavaScript up:

``` ruby
KM.identify("User123")
KM.record("An Event Happened", {"Property 1" => "Value 1", "_t" => 1379025365, "_d" => 1});
```

Upon execution, we would interpret this as:

```
User123 did the event "An Event Happened", at Thu, 12 Sep 2013 22:36:05 GMT.
Wait, we already saw another API call say that this person did this event at the same time. Ignore it.
User123 got the value "Value 1" for the property "Property 1" at Thu, 12 Sep 2013 22:36:05 GMT
Wait, we already saw another API call say that this person got a value for "Property 1" at the same time. Ignore it.
```

This is a basic example of our deduplication in action.

### Example 2

Now, let's say you had the same JavaScript. But instead of the above Ruby, it was actually:

``` ruby
KM.identify("User123")
KM.record("An Event Happened", {"Property 1" => "Value 1", "A Ruby Property" => "A Ruby Value", "_t" => 1379025365, "_d" => 1});
# This is the same, except we're also setting "A Ruby Property"
```

Upon execution, we would interpret this as:

```
User123 did the event "An Event Happened", at Thu, 12 Sep 2013 22:36:05 GMT.
Wait, we already saw another API call say that this person did this event at the same time. Ignore it.
User123 got the value "Value 1" for the property "Property 1" at Thu, 12 Sep 2013 22:36:05 GMT
Wait, we already saw another API call say that this person got a value for "Property 1" at the same time. Ignore it.
User123 got the value "A Ruby Value" for the property "A Ruby Property" at Thu, 12 Sep 2013 22:36:05 GMT
```

In this example, we _did merge_ the results of the JavaScript call with the Ruby call, since the Ruby call provided unique information about setting "A Ruby Value" for "A Ruby Property".

### Example 3

There _is_ a case where clobbering could happen. Let's say you had the same JavaScript, but your Ruby was now:

``` ruby
KM.identify("User123")
KM.record("An Event Happened", {"Property 1" => "Something in Ruby", "_t" => 1379025365, "_d" => 1});
# Notice that the value for "Property 1" is different from what the JS API call provided
```

In this case, we definitely know

```
User123 did the event "An Event Happened", at Thu, 12 Sep 2013 22:36:05 GMT.
```

However, the two API calls are in disagreement as to which value was set for "Property 1". It will be random whether we save the value as "Value 1" or "Something in Ruby".

Does that make things any clearer?

**Comment by steveluscher** [comment]
Most amazing reply ever. So you do what outwardly appears to be a merge of new values into the existing event, except where a value has already been recorded for a given key. Sweet.

By client-y data, I meant things like the screen size and browser version. This is an assumption on my part – I assume that the JS SDK collects and sends that information as a matter of course. If this assumption is false, it's because I haven't actually used KISSMetrics in a while. I'm at the shopping stage, not the using stage.

So, in summary, it _is_ feasible to back client API calls up with server API calls, so long as you use the same timestamp as a correlation factor.

Thanks!

**Comment by eskfung** [comment]
Ah, that's correct. By default, our JavaScript isn't _automatically_ tracking that type of metadata, but it's something you would implement yourself if you needed the data.
